Error when building paseto

**Describe the bug**
When I try to build the latest trunk branch, I get the following error. 

```
error[E0308]: mismatched types
  --> src/v2/local.rs:48:48
   |
48 |   if let Ok(mut state) = GenericHashState::new(24, Some(nonce_key)) {
   |                                                ^^
   |                                                |
   |                                                expected enum `Option`, found integer
   |                                                help: try using a variant of the expected enum: `Some(24)`
   |
   = note: expected enum `Option<usize>`
              found type `{integer}`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: could not compile `paseto`

```

**To Reproduce**
Steps to reproduce the behavior:
1. Check our a fresh copy of the trunk branch or checkout the latest release
2. Run `cargo update` to update compatible dependencies 
3. Run `cargo check`

**Expected behavior**
The crate should compile without error.

**Versions (please complete the following information):**
 - OS: Fedora release 34 (Thirty Four) 5.12.11-300.fc34.x86_64
 - Rustc version rustc 1.53.0 (53cb7b09b 2021-06-17)
 - Version 2.0.1+

**Additional context**

It would appear that sodiumoxide updated their API which caused a breaking change in this library. 

I've forked the repo and will look to provide a PR.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,10 @@
+## 2.0.2+1.0.3
+
+* Migrate away from sodiumoxide as it has become less, and less maintained not
+  providing a stable secure base anymore. Migrate to supported rust-crypto
+  projects. `SodiumErrors` will keep it's name, but now represents these
+  replacements for "sodium".
+
 ## 2.0.1+1.0.3
 
 * Fix mistake where token builder would incorrectly force `nbf`/`iat`/`exp` fields.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,7 +1,7 @@
 [package]
 name = "paseto"
 description = "An alternative token format to JWT"
-version = "2.0.1+1.0.3"
+version = "2.0.2+1.0.3"
 repository = "https://github.com/instructure/paseto"
 license = "MIT"
 authors = [
@@ -12,12 +12,14 @@ edition = "2018"
 [features]
 default = ["v1", "v2", "easy_tokens_chrono"]
 v1 = ["openssl"]
-v2 = ["sodiumoxide"]
+v2 = ["blake2", "chacha20poly1305"]
 easy_tokens_chrono = ["serde_json", "chrono"]
 easy_tokens_time = ["serde_json", "time"]
 
 [dependencies]
 base64 = "^0.13"
+blake2 = { version = "^0.9.1", optional = true }
+chacha20poly1305 = { version = "^0.8.0", optional = true }
 chrono = { version = "^0.4", optional = true, features = ["serde"] }
 time = { version = "^0.2", optional = true, features = ["serde"] }
 failure = "^0.1"
@@ -25,7 +27,6 @@ failure_derive = "^0.1"
 openssl = { version = "~0.10.24", optional = true }
 ring = { version = "^0.16", features = ["std"] }
 serde_json = { version = "^1.0.0", optional = true }
-sodiumoxide = { version = "^0.2.2", optional = true }
 
 [dev-dependencies]
 hex = "^0.4.0"

--- a/src/v2/local.rs
+++ b/src/v2/local.rs
@@ -1,14 +1,24 @@
 //! An Implementation of Paseto V2 "local" tokens (or tokens that are encrypted with a shared secret).
 
-use crate::errors::{GenericError, SodiumErrors};
-use crate::pae::pae;
+use crate::{
+  errors::{GenericError, SodiumErrors},
+  pae::pae,
+};
 
 use base64::{decode_config, encode_config, URL_SAFE_NO_PAD};
+use blake2::{
+  digest::{Update, VariableOutput},
+  VarBlake2b,
+};
+use chacha20poly1305::{
+  aead::{Aead, NewAead, Payload},
+  XChaCha20Poly1305, XNonce,
+};
 use failure::Error;
-use ring::constant_time::verify_slices_are_equal as ConstantTimeEquals;
-use ring::rand::{SecureRandom, SystemRandom};
-use sodiumoxide::crypto::aead::xchacha20poly1305_ietf::{open as Decrypt, seal as Encrypt, Key, Nonce};
-use sodiumoxide::crypto::generichash::State as GenericHashState;
+use ring::{
+  constant_time::verify_slices_are_equal as ConstantTimeEquals,
+  rand::{SecureRandom, SystemRandom},
+};
 
 const HEADER: &str = "v2.local.";
 
@@ -44,49 +54,43 @@ pub fn local_paseto(msg: &str, footer: Option<&str>, key: &[u8]) -> Result<Strin
 /// `key` - The key to encrypt the message with.
 fn underlying_local_paseto(msg: &str, footer: Option<&str>, nonce_key: &[u8; 24], key: &[u8]) -> Result<String, Error> {
   let footer_frd = footer.unwrap_or("");
+  let mut state = VarBlake2b::new_keyed(nonce_key, 24);
+  state.update(msg.as_bytes());
+  let finalized = state.finalize_boxed();
+  let nonce = XNonce::from_slice(finalized.as_ref());
+  if let Ok(aead) = XChaCha20Poly1305::new_from_slice(key) {
+    let pre_auth = pae(&[HEADER.as_bytes(), finalized.as_ref(), footer_frd.as_bytes()]);
+    if let Ok(crypted) = aead.encrypt(
+      nonce,
+      Payload {
+        msg: msg.as_bytes(),
+        aad: pre_auth.as_ref(),
+      },
+    ) {
+      let mut n_and_c = Vec::new();
+      n_and_c.extend_from_slice(finalized.as_ref());
+      n_and_c.extend_from_slice(crypted.as_ref());
 
-  if let Ok(mut state) = GenericHashState::new(24, Some(nonce_key)) {
-    if let Ok(_) = state.update(msg.as_bytes()) {
-      if let Ok(finalized) = state.finalize() {
-        let nonce_bytes = finalized.as_ref();
-        if let Some(nonce) = Nonce::from_slice(nonce_bytes) {
-          let key_obj = Key::from_slice(key);
-          if key_obj.is_none() {
-            return Err(SodiumErrors::InvalidKey {})?;
-          }
-          let key_obj = key_obj.unwrap();
-
-          let pre_auth = pae(&[HEADER.as_bytes(), &nonce_bytes, footer_frd.as_bytes()]);
-
-          let crypted = Encrypt(msg.as_bytes(), Some(pre_auth.as_ref()), &nonce, &key_obj);
-
-          let mut n_and_c = Vec::new();
-          n_and_c.extend_from_slice(&nonce_bytes);
-          n_and_c.extend_from_slice(&crypted);
-
-          let token = if !footer_frd.is_empty() {
-            format!(
-              "{}{}.{}",
-              HEADER,
-              encode_config(&n_and_c, URL_SAFE_NO_PAD),
-              encode_config(footer_frd.as_bytes(), URL_SAFE_NO_PAD)
-            )
-          } else {
-            format!("{}{}", HEADER, encode_config(&n_and_c, URL_SAFE_NO_PAD))
-          };
-
-          Ok(token)
-        } else {
-          Err(SodiumErrors::FunctionError {})?
-        }
+      let token = if !footer_frd.is_empty() {
+        format!(
+          "{}{}.{}",
+          HEADER,
+          encode_config(&n_and_c, URL_SAFE_NO_PAD),
+          encode_config(footer_frd.as_bytes(), URL_SAFE_NO_PAD)
+        )
       } else {
-        Err(SodiumErrors::FunctionError {})?
-      }
+        format!("{}{}", HEADER, encode_config(&n_and_c, URL_SAFE_NO_PAD))
+      };
+
+      Ok(token)
     } else {
-      Err(SodiumErrors::FunctionError {})?
+      return Err(SodiumErrors::FunctionError {})?;
     }
   } else {
-    Err(SodiumErrors::FunctionError {})?
+    return Err(SodiumErrors::InvalidKeySize {
+      size_provided: key.len(),
+      size_needed: 32,
+    })?;
   }
 }
 
@@ -124,21 +128,24 @@ pub fn decrypt_paseto(token: &str, footer: Option<&str>, key: &[u8]) -> Result<S
 
   let pre_auth = pae(&[HEADER.as_bytes(), nonce, footer_str.as_bytes()]);
 
-  let nonce_obj = Nonce::from_slice(nonce);
-  let key_obj = Key::from_slice(key);
-  if nonce_obj.is_none() || key_obj.is_none() {
+  let nonce_obj = XNonce::from_slice(nonce);
+  let aead = XChaCha20Poly1305::new_from_slice(key);
+  if aead.is_err() {
     return Err(SodiumErrors::InvalidKey {})?;
   }
-  let nonce_obj = nonce_obj.unwrap();
-  let key_obj = key_obj.unwrap();
-
-  let decrypted = Decrypt(ciphertext, Some(&pre_auth), &nonce_obj, &key_obj);
+  let aead = aead.unwrap();
+  let decrypted = aead.decrypt(
+    nonce_obj,
+    Payload {
+      msg: ciphertext,
+      aad: &pre_auth,
+    },
+  );
   if decrypted.is_err() {
     return Err(SodiumErrors::FunctionError {})?;
   }
-  let decrypted = decrypted.unwrap();
 
-  Ok(String::from_utf8(decrypted)?)
+  Ok(String::from_utf8(decrypted.unwrap())?)
 }
 
 #[cfg(test)]

